File: VB/Task26/MyDateEditCalendar.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Windows.Forms
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.XtraEditors.Calendar
Imports DevExpress.Utils.Menu
Imports DevExpress.Utils

Namespace Task26
	Public Class MyDateEditCalendar
		Inherits DateEditCalendar
		Private fLastHitInfoDate As DateTime
		Private fNoteMenu As ContextMenu

		Public Sub New(ByVal item As RepositoryItemMyDateEdit, ByVal editDate As Object)
			MyBase.New(item, editDate)
			fLastHitInfoDate = New DateTime()
			fNoteMenu = CreateNoteMenu()
		End Sub

		Public Shadows ReadOnly Property Properties() As RepositoryItemMyDateEdit
			Get
				Return TryCast(MyBase.Properties, RepositoryItemMyDateEdit)
			End Get
		End Property

		Protected Overrides Function CreatePainter() As DateEditPainter
			Return New MyDateEditCalendarPainter(Me)
		End Function

		Protected Overridable ReadOnly Property TipController() As ToolTipController
			Get
				Return ToolTipController.DefaultController
			End Get
		End Property

		Protected Overrides Sub OnMouseUp(ByVal e As MouseEventArgs)
			If e.Button = MouseButtons.Right Then
				Dim HI As CalendarHitInfo = GetHitInfo(e)
				If HI.InfoType = CalendarHitInfoType.MonthNumber Then
					Me.DateTime = HI.HitDate
					Dim NotEmpty As Boolean = Properties.HaveNote(HI.HitDate)
					fNoteMenu.MenuItems(1).Enabled = NotEmpty
					fNoteMenu.MenuItems(2).Enabled = NotEmpty
					fNoteMenu.Show(Me, e.Location)
				End If
			Else
				MyBase.OnMouseUp(e)
			End If
		End Sub

		Protected Overrides Sub OnMouseMove(ByVal e As MouseEventArgs)
			MyBase.OnMouseMove(e)
			Dim HI As CalendarHitInfo = GetHitInfo(e)
			If HI.InfoType = CalendarHitInfoType.MonthNumber Then
				If HI.HitDate <> fLastHitInfoDate Then
					fLastHitInfoDate = HI.HitDate
					Dim Notes As List(Of Note) = Properties.GetDateNotes(HI.HitDate)
					If Notes.Count > 0 Then
						Dim NotesList As String = ""
						For Each N As Note In Notes
							NotesList &= N.Date.ToShortTimeString() & ": " & N.Name + ControlChars.Lf
						Next N
						TipController.ShowHint(NotesList, HI.HitDate.ToShortDateString() & ": " & Notes.Count & " Notes.", PointToScreen(e.Location))
						Return
					Else
						TipController.HideHint()
					End If
				End If
			Else
				TipController.HideHint()
			End If
		End Sub

		Protected Overridable Function CreateNoteMenu() As ContextMenu
			Dim CM As New ContextMenu()
			CM.MenuItems.Add(New MenuItem("Add Note...", New EventHandler(AddressOf OnMenuAddNote_Click)))

			CM.MenuItems.Add(New MenuItem("Edit...", New EventHandler(AddressOf OnMenuEditNote_Click)))
			CM.MenuItems.Add(New MenuItem("Delete...", New EventHandler(AddressOf OnMenuDeleteNote_Click)))

			Return CM
		End Function

		Protected Overridable Sub OnMenuAddNote_Click(ByVal sender As Object, ByVal e As EventArgs)
			Properties.AddNote(DateTime)
		End Sub

		Protected Overridable Sub OnMenuEditNote_Click(ByVal sender As Object, ByVal e As EventArgs)
			Properties.EditNotes(DateTime)
		End Sub

		Protected Overridable Sub OnMenuDeleteNote_Click(ByVal sender As Object, ByVal e As EventArgs)
			Properties.DeleteNotes(DateTime)
		End Sub
	End Class
End Namespace

File: CS/Task26/MyDateEditCalendar.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors.Controls;
using DevExpress.XtraEditors.Calendar;
using DevExpress.Utils.Menu;
using DevExpress.Utils;

namespace Task26
{
    public class MyDateEditCalendar : DateEditCalendar
    {
        DateTime fLastHitInfoDate;
        ContextMenu fNoteMenu;
        
        public MyDateEditCalendar(RepositoryItemMyDateEdit item, object editDate)
            : base(item, editDate)
        {
            fLastHitInfoDate = new DateTime();
            fNoteMenu = CreateNoteMenu();
        }

        public new RepositoryItemMyDateEdit Properties
        {
            get { return base.Properties as RepositoryItemMyDateEdit; }
        }

        protected override DateEditPainter CreatePainter()
        {
            return new MyDateEditCalendarPainter(this);
        }

        protected virtual ToolTipController TipController
        {
            get { return ToolTipController.DefaultController; }
        }

        protected override void OnMouseUp(MouseEventArgs e)
        {
            if (e.Button == MouseButtons.Right)
            {
                CalendarHitInfo HI = GetHitInfo(e);
                if (HI.InfoType == CalendarHitInfoType.MonthNumber)
                {
                    this.DateTime = HI.HitDate;
                    bool NotEmpty = Properties.HaveNote(HI.HitDate);
                    fNoteMenu.MenuItems[1].Enabled = NotEmpty;
                    fNoteMenu.MenuItems[2].Enabled = NotEmpty;
                    fNoteMenu.Show(this, e.Location);
                }
            }
            else
                base.OnMouseUp(e);
        }

        protected override void OnMouseMove(MouseEventArgs e)
        {
            base.OnMouseMove(e);
            CalendarHitInfo HI = GetHitInfo(e);
            if (HI.InfoType == CalendarHitInfoType.MonthNumber)
            {
                if (HI.HitDate != fLastHitInfoDate)
                {
                    fLastHitInfoDate = HI.HitDate;
                    List<Note> Notes = Properties.GetDateNotes(HI.HitDate);
                    if (Notes.Count > 0)
                    {
                        string NotesList = "";
                        foreach (Note N in Notes)
                        {
                            NotesList += N.Date.ToShortTimeString() + ": " + N.Name + '\n';
                        }
                        TipController.ShowHint(NotesList, HI.HitDate.ToShortDateString() + ": " + Notes.Count + " Notes.", PointToScreen(e.Location));
                        return;
                    }
                    else
                        TipController.HideHint();
                }
            }
            else
                TipController.HideHint();
        }
        
        protected virtual ContextMenu CreateNoteMenu()
        {
            ContextMenu CM = new ContextMenu();
            CM.MenuItems.Add(new MenuItem("Add Note...", new EventHandler(OnMenuAddNote_Click)));
           
            CM.MenuItems.Add(new MenuItem("Edit...", new EventHandler(OnMenuEditNote_Click)));
            CM.MenuItems.Add(new MenuItem("Delete...", new EventHandler(OnMenuDeleteNote_Click)));

            return CM;
        }

        protected virtual void OnMenuAddNote_Click(object sender, EventArgs e)
        {
            Properties.AddNote(DateTime);
        }

        protected virtual void OnMenuEditNote_Click(object sender, EventArgs e)
        {
            Properties.EditNotes(DateTime);
        }

        protected virtual void OnMenuDeleteNote_Click(object sender, EventArgs e)
        {
            Properties.DeleteNotes(DateTime);
        }
    }
}
